docs/custom_promise_types/groups.py

```diff
@@ -0,0 +1,338 @@
+            
+            # make sure users aren't both included and excluded
+            if "include" in members and "exclude" in members:
+                include = members["include"]
+                exclude = members["exclude"]
+                for included in include:
+                    for excluded in exclude:
+                        if included == excluded:
+                            raise ValidationError(f"user '{included}' both included and excluded from group '{promiser}'")
+
+
+    def evaluate_promise(self, promiser, attributes):
+        # keep track of any repairs or failed repairs
+        failed_repairs = 0
+        repairs = 0
+
+        # get group from '/etc/group' if present
+        group = None
+        try:
+            group = Group.lookup(promiser)
+        except GroupException as e:
+            self.log_error(f"failed to lookup group '{promiser}': {e}")
+            failed_repairs += 1
+        
+        # get promised gid if present
+        promised_gid = attributes["gid"] if "gid" in attributes else None
+
+        # parse json in attribute members
+        if "members" in attributes:
+            # remove escape chars followed by parsing json
+            members = json.loads(json.loads('"' + attributes["members"] + '"'))
+            attributes["members"] = members
+
+        # set policy to present by default, if not specified
+        if "policy" not in attributes:
+            self.log_verbose(f"policy not specified, defaults to present")
+            attributes["policy"] = "present"
+
+        # create group if policy present and group absent
+        if attributes["policy"] == "present" and group is None:
+            self.log_debug(f"group '{promiser}' should be present, but does not exist")
+            try:
+                self.log_info(f"creating group '{promiser}'")
+                group = Group.create(promiser, promised_gid)
+                repairs += 1
+            except GroupException as e:
+                self.log_error(f"failed to create group '{promiser}': {e}")
+                failed_repairs += 1
+
+        # delete group if policy absent and group present
+        if attributes["policy"] == "absent" and group is not None:
+            self.log_debug(f"group '{promiser}' should be absent, but does exist")
+            try:
+                self.log_info(f"deleting group '{promiser}'")
+                group = group.delete()
+                repairs += 1
+            except GroupException as e:
+                self.log_error(f"failed to delete group '{promiser}': {e}")
+                failed_repairs += 1
+
+        # if group is now present, check attributes 'gid' and 'members'
+        if group:
+            # check gid if present
+            if promised_gid and promised_gid != group.gid:
+                self.log_error(f"GID '{group.gid}' for promiser '{promiser}' does not match promised GID '{promised_gid}'")
```

Comment:
- Yes, if we failed to create group with GID, its probably do to the fact that an existing group already has that GID. 
 - If group already exist, but with a different GID, it does not imply that an existing group has that GID. Thus  "There is an existing group ..."  may be a false statement.

Reply:
>     * Yes, if we failed to create group with GID, its probably do to the fact that an existing group already has that GID.

But that should be handled above where `Group.create()` is called.


>     * If group already exist, but with a different GID, it does not imply that an existing group has that GID.

`Group.lookup()` searches by name, right? So `group.gid` would be GID of an existing group with the same name here.

Am I missing something?

Reply:
It compares the promiser's GID to the promised GID. Thus if the promiser's GID does not match the GID it was promised, it does not imply that there is another group with that GID. It only means the promiser's GID did not match the promised GID.

This can happen if the promiser already exists, but with another GID. The error message will be: `"GID '123' for promiser 'foo' does not match promised GID '321'"`

If another group has the promised GID and the promiser does not exist. Creating the promiser will fail. And the error message will be: `"failed to create group 'foo': groupadd: group 'foo' already exists"`

The `lookup` function is used to look up the promiser group.

Reply:
I do understand that. It's just that `"GID '123' for promiser 'foo' does not match promised GID '321'"` is, for me, a very cryptic way of saying *There's an existing group foo which has a different GID ('123') than what was promised ('321')*.

Reply:
OK, then I understand. My bad. I thought you meant "there is another group than `foo`".
I agreee `"There's an existing group foo which has a different GID ('123') than what was promised ('321')"` is a lot clearer.